### PR TITLE
[LI-TIERING] Added default RLMM implementation based on internal topic storage.

### PR DESCRIPTION
**Pull request by wyuka**
Kafka JIRA: [KAFKA-9555](https://issues.apache.org/jira/browse/KAFKA-9555)
Source commit: https://github.com/satishd/kafka/commit/25f9120ddf4bb2713c3b4d89cc31fcb411aeba1a

Summary:
 - Introduced TopicBasedRemoteLogMetadataManagerHarness which takes care of bringing up a Kafka cluster and create remote log metadata topic and initializes TopicBasedRemoteLogMetadataManager.
 - Added TopicBasedRemoteLogMetadataManager to run the existing tests by making the existing RemoteLogMetadataManagerTest parameterized.
 - Added transformation of TopicIdPartition into bytes with custom hashing.
 - Skipping the events which are not currently assigned for the ConsumerTask.

Added file based remote log metadata snapshots:
 - Added TopicBasedRemoteLogMetadataManagerRestartTest and fixed issues in reloading the snapshots.
 - TopicBasedRemoteLogMetadataManagerRestartTest is about verifying:
      * load the earlier saved snapshots after restart
      * check the entries are available
      * start the consumer and add more metadata entries
      * check the newly addded entries and loaded entries are available
 - Fixed a few issues in snapshot read/write/load and added tests.

This is part of tiered storage KIP-405 efforts.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by smccauliff** [comment]
Did you make any modifications to the original code?

**Review by smccauliff — APPROVED** [review]
I didn't read every line of code here, but at some point I've looked at the core RLMM implementation and didn't find anything that definitely needed to be fixed.  I'm going to give +1 for this.  

@gitlw Do you want to spend more time looking at this change?  It seems almost completely orthogonal to the normal non-tiered code path.

File: build.gradle [hunk 1]
```diff
@@ -1555,8 +1555,12 @@ project(':storage') {
 
     testImplementation project(':clients')
     testImplementation project(':clients').sourceSets.test.output
+    testImplementation project(':core')
```
Comment (wyuka) [inline build.gradle:1559]: Changed `testCompile` to `testImplementation`, because `testCompile` was deprecated, and is now removed in Gradle 7.0.

**Review by wyuka — COMMENTED** [review]
The only changes made on top of the source commit are to `build.gradle`.

File: storage/src/main/resources/message/RemotePartitionDleteMetadataSnapshot.json [hunk 2]
```diff
@@ -0,0 +1,42 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
```
Comment (gitlw) [inline storage/src/main/resources/message/RemotePartitionDleteMetadataSnapshot.json]: There is a typo in this file name. Dlete -> Delete.
Does it make sense to correct it or leave the typo as it is?

**Review by gitlw — APPROVED** [review]
Please check my minor comment and see if that needs to be fixed.
Otherwise this PR LGTM. Thanks!

File: storage/src/main/resources/message/RemotePartitionDleteMetadataSnapshot.json (hunk 2 above)
Comment (wyuka) [reply storage/src/main/resources/message/RemotePartitionDleteMetadataSnapshot.json]: I think we can change the name. I don't see any problems with doing that.

**Pull request merged**